### ISSUE TITLE
[JENKINS-63785] Git plugin uses wrong branch unless second fetch is preserved

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/git-plugin#3605]
<p>Hello,</p>

<p>First, thank you for maintaining and developing this nice Git plugin!</p>

<p>Recently, I got some issues with some jobs using a wrong refspec to build. This only happened recently for jobs doing a "frech clone" and I suspect, only after an update of the Git Plugin.</p>

<p>I don't have the issue when enabling "<a href="https://plugins.jenkins.io/git/#global-configuration" class="external-link" target="_blank" rel="nofollow noopener">Preserve second fetch during checkout</a>" option from PR-845 (<a href="https://issues.jenkins.io/browse/JENKINS-49757" title="Git plugin calls fetch twice per checkout" class="issue-link" data-issue-key="JENKINS-49757"><del>JENKINS-49757</del></a>).</p>

<p>Maybe what is particular on my side is that there are multiple branches in my repo and my refspec contains '<tt>refs/heads</tt>' ("<tt>refs/heads/mptcp_trunk/master"</tt>). From what I see in PR-845, that's why the second fetch is not done but I don't know why it is an issue to add '<tt>refs/heads</tt>' in the refspec. When I look at what is done by Jenkins, it looks like I could only add the name of the branch but if I have not to specified '<tt>refs/heads</tt>', I am confused with the option name and the description <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
 Also, I didn't set anything related to "<em>Advanced Clone Option</em>" as mentioned in the doc because I didn't find the corresponding option in the config (and to be honest, I don't know why I would like to do the clone and the fetch later differently <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ).</p>

<p>(Note: I added "origin/" in the branch name just to try but it didn't help. Just to explain why we see a reference to "origin/xxx" in the logs)</p>

<p>Here are the logs without "Preserve second fetch during checkout" option:</p>

<pre>
 > git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git +refs/heads/*:refs/remotes/origin/* # timeout=10
</span> > git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span> > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
Avoid second fetch
 > git rev-parse FETCH_HEAD^{commit} # timeout=10
Checking out Revision 5ac0ebcbb890f2a7852bca357641163837a219bd (origin/mptcp_trunk/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 5ac0ebcbb890f2a7852bca357641163837a219bd # timeout=10
Commit message: <span class="code-quote">"old commit"</span>
</pre>
<p>Here, you can see that all branches are fetches but then <tt>FETCH_HEAD</tt> is used with <tt>git checkout</tt> which is not correct: <tt>FETCH_HEAD</tt> corresponds to the first branch by alphabetical order on the remote (because we asked for <tt>refs/heads/*</tt>).</p>

<p>Now here is the output with "Preserve second fetch during checkout" on:</p>

<pre>
 > git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git +refs/heads/*:refs/remotes/origin/* # timeout=10
</span> > git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span> > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 > git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span>Fetching upstream changes from ssh:<span class="code-comment">//gerrit:29418/repo.git
</span>using GIT_SSH to set credentials 
 > git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git refs/heads/mptcp_trunk/master +refs/heads/mptcp_trunk/master:refs/remotes/origin/mptcp_trunk/master # timeout=10
</span> > git rev-parse FETCH_HEAD^{commit} # timeout=10
Checking out Revision 21ec2388799b5bd7e18f3f2bd4aa7f0e2be9bb23 (origin/mptcp_trunk/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 21ec2388799b5bd7e18f3f2bd4aa7f0e2be9bb23 # timeout=10
Commit message: <span class="code-quote">"<span class="code-keyword">new</span> commit"</span>
</pre>
<p><tt>FETCH_HEAD</tt> is now corresponding to <tt>refs/heads/mptcp_trunk/master</tt>, what I want: then everything is OK.</p>

<p>I hope this will help finding the issue not to have to enable "Preserve second fetch during checkout" option.</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matttbe.png" title="matttbe's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matttbe">matttbe</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785" target="_blank">Git plugin uses wrong branch unless second fetch is preserved</a></i></summary>
<i><ul>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: git-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2025-04-06T19:48:01+00:00
<li><b>votes</b>: 2
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-02
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>Hello,</p>

<p>First, thank you for maintaining and developing this nice Git plugin!</p>

<p>Recently, I got some issues with some jobs using a wrong refspec to build. This only happened recently for jobs doing a "frech clone" and I suspect, only after an update of the Git Plugin.</p>

<p>I don't have the issue when enabling "<a href="https://plugins.jenkins.io/git/#global-configuration" class="external-link" target="_blank" rel="nofollow noopener">Preserve second fetch during checkout</a>" option from PR-845 (<a href="https://issues.jenkins.io/browse/JENKINS-49757" title="Git plugin calls fetch twice per checkout" class="issue-link" data-issue-key="JENKINS-49757"><del>JENKINS-49757</del></a>).</p>

<p>Maybe what is particular on my side is that there are multiple branches in my repo and my refspec contains '<tt>refs/heads</tt>' ("<tt>refs/heads/mptcp_trunk/master"</tt>). From what I see in PR-845, that's why the second fetch is not done but I don't know why it is an issue to add '<tt>refs/heads</tt>' in the refspec. When I look at what is done by Jenkins, it looks like I could only add the name of the branch but if I have not to specified '<tt>refs/heads</tt>', I am confused with the option name and the description <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
 Also, I didn't set anything related to "<em>Advanced Clone Option</em>" as mentioned in the doc because I didn't find the corresponding option in the config (and to be honest, I don't know why I would like to do the clone and the fetch later differently <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ).</p>

<p>(Note: I added "origin/" in the branch name just to try but it didn't help. Just to explain why we see a reference to "origin/xxx" in the logs)</p>

<p>Here are the logs without "Preserve second fetch during checkout" option:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"> &gt; git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git +refs/heads/*:refs/remotes/origin/* # timeout=10
</span> &gt; git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span> &gt; git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
Avoid second fetch
 &gt; git rev-parse FETCH_HEAD^{commit} # timeout=10
Checking out Revision 5ac0ebcbb890f2a7852bca357641163837a219bd (origin/mptcp_trunk/master)
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 5ac0ebcbb890f2a7852bca357641163837a219bd # timeout=10
Commit message: <span class="code-quote">"old commit"</span>
</pre>
</div></div>
<p>Here, you can see that all branches are fetches but then <tt>FETCH_HEAD</tt> is used with <tt>git checkout</tt> which is not correct: <tt>FETCH_HEAD</tt> corresponds to the first branch by alphabetical order on the remote (because we asked for <tt>refs/heads/*</tt>).</p>

<p>Now here is the output with "Preserve second fetch during checkout" on:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"> &gt; git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git +refs/heads/*:refs/remotes/origin/* # timeout=10
</span> &gt; git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span> &gt; git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
 &gt; git config remote.origin.url ssh:<span class="code-comment">//gerrit:29418/repo.git # timeout=10
</span>Fetching upstream changes from ssh:<span class="code-comment">//gerrit:29418/repo.git
</span>using GIT_SSH to set credentials 
 &gt; git fetch --tags --progress ssh:<span class="code-comment">//gerrit:29418/repo.git refs/heads/mptcp_trunk/master +refs/heads/mptcp_trunk/master:refs/remotes/origin/mptcp_trunk/master # timeout=10
</span> &gt; git rev-parse FETCH_HEAD^{commit} # timeout=10
Checking out Revision 21ec2388799b5bd7e18f3f2bd4aa7f0e2be9bb23 (origin/mptcp_trunk/master)
 &gt; git config core.sparsecheckout # timeout=10
 &gt; git checkout -f 21ec2388799b5bd7e18f3f2bd4aa7f0e2be9bb23 # timeout=10
Commit message: <span class="code-quote">"<span class="code-keyword">new</span> commit"</span>
</pre>
</div></div>
<p><tt>FETCH_HEAD</tt> is now corresponding to <tt>refs/heads/mptcp_trunk/master</tt>, what I want: then everything is OK.</p>

<p>I hope this will help finding the issue not to have to enable "Preserve second fetch during checkout" option.</p></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Git Plugin 4.4.3<br/>
Jenkins 2.249.1<br/>
CentOS 7
```
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=398155&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398155">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for reporting the issue.  It looks like you've found a case that is missed in our logic to decide if the second fetch is required.  Could you upload the config.xml for your job definition and / or the <tt>checkout</tt> step that is being performed (if this is a Pipeline job).</p>

<p>I suspect that the job is configured with multiple repositories or with multiple refspecs and the plugin is not handling that correctly.  We'll need the job details to assure that we can duplicate the problem and provide a fix.  In this case, I expect that there is a way that the job can be redefined so that it does not require the second fetch.  Whether there is or not, this is a case where the second fetch is required and the plugin did not detect that the second fetch is required.  Sorry we missed that!</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for reporting the issue.  It looks like you've found a case that is missed in our logic to decide if the second fetch is required.  Could you upload the config.xml for your job definition and / or the <tt>checkout</tt> step that is being performed (if this is a Pipeline job).</p>

<p>I suspect that the job is configured with multiple repositories or with multiple refspecs and the plugin is not handling that correctly.  We'll need the job details to assure that we can duplicate the problem and provide a fix.  In this case, I expect that there is a way that the job can be redefined so that it does not require the second fetch.  Whether there is or not, this is a case where the second fetch is required and the plugin did not detect that the second fetch is required.  Sorry we missed that!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matttbe.png" title="matttbe's avatar" /> matttbe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=398159&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398159">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Mark,</p>

<p>Thank you for your quick reply!</p>
<blockquote><p>Could you upload the config.xml for your job definition and / or the <tt>checkout</tt> step that is being performed (if this is a Pipeline job).</p>

<p>I suspect that the job is configured with multiple repositories or with multiple refspecs and the plugin is not handling that correctly. We'll need the job details to assure that we can duplicate the problem and provide a fix.</p></blockquote>
<p>This is not a Pipeline job, there is no multiple repositories but multiple refspecs:</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">  &lt;scmCheckoutRetryCount&gt;3&lt;/scmCheckoutRetryCount&gt;
  &lt;properties&gt;
    &lt;hudson.model.ParametersDefinitionProperty&gt;
      &lt;parameterDefinitions&gt;
        &lt;hudson.model.StringParameterDefinition&gt;
          &lt;name&gt;GERRIT_PROJECT&lt;/name&gt;
          &lt;description&gt;GERRIT_PROJECT parameter <span class="code-keyword">if</span> not given by trigger.&lt;/description&gt;
          &lt;defaultValue&gt;repo&lt;/defaultValue&gt;
        &lt;/hudson.model.StringParameterDefinition&gt;
        &lt;hudson.model.StringParameterDefinition&gt;
          &lt;name&gt;GERRIT_BRANCH&lt;/name&gt;
          &lt;description&gt;JJB configured GERRIT_BRANCH parameter.&lt;/description&gt;
          &lt;defaultValue&gt;mptcp_trunk/master&lt;/defaultValue&gt;
        &lt;/hudson.model.StringParameterDefinition&gt;
        &lt;hudson.model.StringParameterDefinition&gt;
          &lt;name&gt;GERRIT_REFSPEC&lt;/name&gt;
          &lt;description&gt;GERRIT_REFSPEC parameter <span class="code-keyword">if</span> not given by trigger. DO NOT CHANGE THIS: <span class="code-keyword">if</span> the build triggered by <span class="code-quote">"pollscm"</span>: pollscm will follow the last refspec modified here instead of the <span class="code-keyword">default</span> value.&lt;/description&gt;
          &lt;defaultValue&gt;refs/heads/mptcp_trunk/master&lt;/defaultValue&gt;
        &lt;/hudson.model.StringParameterDefinition&gt;
      &lt;/parameterDefinitions&gt;
    &lt;/hudson.model.ParametersDefinitionProperty&gt;
  &lt;/properties&gt;
  &lt;scm class=<span class="code-quote">"hudson.plugins.git.GitSCM"</span>&gt;
    &lt;configVersion&gt;2&lt;/configVersion&gt;
    &lt;userRemoteConfigs&gt;
      &lt;hudson.plugins.git.UserRemoteConfig&gt;
        &lt;name&gt;origin&lt;/name&gt;
        &lt;refspec&gt;$GERRIT_REFSPEC +refs/heads/$GERRIT_BRANCH:refs/remotes/origin/$GERRIT_BRANCH&lt;/refspec&gt;
        &lt;url&gt;ssh:<span class="code-comment">//gerrit:29418/repo.git&lt;/url&gt;
</span>        &lt;credentialsId&gt;REMOVED&lt;/credentialsId&gt;
      &lt;/hudson.plugins.git.UserRemoteConfig&gt;
    &lt;/userRemoteConfigs&gt;
    &lt;branches&gt;
      &lt;hudson.plugins.git.BranchSpec&gt;
        &lt;name&gt;origin/$GERRIT_BRANCH&lt;/name&gt;
      &lt;/hudson.plugins.git.BranchSpec&gt;
    &lt;/branches&gt;
    &lt;doGenerateSubmoduleConfigurations&gt;<span class="code-keyword">false</span>&lt;/doGenerateSubmoduleConfigurations&gt;
    &lt;remotePoll&gt;<span class="code-keyword">false</span>&lt;/remotePoll&gt;
    &lt;gitTool&gt;Default&lt;/gitTool&gt;
    &lt;submoduleCfg class=<span class="code-quote">"list"</span>/&gt;
    &lt;reference/&gt;
    &lt;gitConfigName&gt;NAME&lt;/gitConfigName&gt;
    &lt;gitConfigEmail&gt;EMAIL&lt;/gitConfigEmail&gt;
    &lt;extensions&gt;
      &lt;hudson.plugins.git.extensions.impl.BuildChooserSetting&gt;
        &lt;buildChooser class=<span class="code-quote">"com.sonyericsson.hudson.plugins.gerrit.trigger.hudsontrigger.GerritTriggerBuildChooser"</span>/&gt;
      &lt;/hudson.plugins.git.extensions.impl.BuildChooserSetting&gt;
      &lt;hudson.plugins.git.extensions.impl.SubmoduleOption&gt;
        &lt;disableSubmodules&gt;<span class="code-keyword">true</span>&lt;/disableSubmodules&gt;
        &lt;recursiveSubmodules&gt;<span class="code-keyword">false</span>&lt;/recursiveSubmodules&gt;
        &lt;trackingSubmodules&gt;<span class="code-keyword">false</span>&lt;/trackingSubmodules&gt;
        &lt;parentCredentials&gt;<span class="code-keyword">false</span>&lt;/parentCredentials&gt;
        &lt;reference/&gt;
        &lt;timeout&gt;10&lt;/timeout&gt;
      &lt;/hudson.plugins.git.extensions.impl.SubmoduleOption&gt;
    &lt;/extensions&gt;
    &lt;browser class=<span class="code-quote">"hudson.plugins.git.browser.GitWeb"</span>&gt;
      &lt;url&gt;https:<span class="code-comment">//gerrit/gitweb?p=repo.git&lt;/url&gt;
</span>    &lt;/browser&gt;
  &lt;/scm&gt;
</pre>
</div></div>
<p>Do you need more than the properties and scm sections? (If yes, I might have to strip some info: emails address, scripts, etc. because this example is coming from a "private" instance, I didn't have issues with OpenSource projects because they didn't do a clone recently <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> )</p>
<blockquote><p>In this case, I expect that there is a way that the job can be redefined so that it does not require the second fetch. Whether there is or not, this is a case where the second fetch is required and the plugin did not detect that the second fetch is required.</p></blockquote>
<p>In my case, the first fetch getting all branches is not needed. It doesn't really hurt but I don't need all branches.</p>

<p>But if for some reasons, all branches have to be fetched, a second fetch to the same repo is not needed if the plugin can use the first refspec in the config list instead of FETCH_HEAD. But I don't know if other users rely on FETCH_HEAD later. I guess they should use HEAD instead anyway.</p>
<blockquote><p>Sorry we missed that!</p></blockquote>
<p>All good, this happens <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Cheers,
Matt</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi Mark,</p>

<p>Thank you for your quick reply!</p>
<blockquote><p>Could you upload the config.xml for your job definition and / or the <tt>checkout</tt> step that is being performed (if this is a Pipeline job).</p>

<p>I suspect that the job is configured with multiple repositories or with multiple refspecs and the plugin is not handling that correctly. We'll need the job details to assure that we can duplicate the problem and provide a fix.</p></blockquote>
<p>This is not a Pipeline job, there is no multiple repositories but multiple refspecs:</p>

<p> </p>

<pre>
  <scmCheckoutRetryCount>3</scmCheckoutRetryCount>
  <properties>
    <hudson.model.ParametersDefinitionProperty>
      <parameterDefinitions>
<hudson.model.StringParameterDefinition>
  <name>GERRIT_PROJECT</name>
  <description>GERRIT_PROJECT parameter <span class="code-keyword">if</span> not given by trigger.</description>
  <defaultValue>repo</defaultValue>
</hudson.model.StringParameterDefinition>
<hudson.model.StringParameterDefinition>
  <name>GERRIT_BRANCH</name>
  <description>JJB configured GERRIT_BRANCH parameter.</description>
  <defaultValue>mptcp_trunk/master</defaultValue>
</hudson.model.StringParameterDefinition>
<hudson.model.StringParameterDefinition>
  <name>GERRIT_REFSPEC</name>
  <description>GERRIT_REFSPEC parameter <span class="code-keyword">if</span> not given by trigger. DO NOT CHANGE THIS: <span class="code-keyword">if</span> the build triggered by <span class="code-quote">"pollscm"</span>: pollscm will follow the last refspec modified here instead of the <span class="code-keyword">default</span> value.</description>
  <defaultValue>refs/heads/mptcp_trunk/master</defaultValue>
</hudson.model.StringParameterDefinition>
      </parameterDefinitions>
    </hudson.model.ParametersDefinitionProperty>
  </properties>
  <scm class=<span class="code-quote">"hudson.plugins.git.GitSCM"</span>>
    <configVersion>2</configVersion>
    <userRemoteConfigs>
      <hudson.plugins.git.UserRemoteConfig>
<name>origin</name>
<refspec>$GERRIT_REFSPEC +refs/heads/$GERRIT_BRANCH:refs/remotes/origin/$GERRIT_BRANCH</refspec>
<url>ssh:<span class="code-comment">//gerrit:29418/repo.git</url>
</span><credentialsId>REMOVED</credentialsId>
      </hudson.plugins.git.UserRemoteConfig>
    </userRemoteConfigs>
    <branches>
      <hudson.plugins.git.BranchSpec>
<name>origin/$GERRIT_BRANCH</name>
      </hudson.plugins.git.BranchSpec>
    </branches>
    <doGenerateSubmoduleConfigurations><span class="code-keyword">false</span></doGenerateSubmoduleConfigurations>
    <remotePoll><span class="code-keyword">false</span></remotePoll>
    <gitTool>Default</gitTool>
    <submoduleCfg class=<span class="code-quote">"list"</span>/>
    <reference/>
    <gitConfigName>NAME</gitConfigName>
    <gitConfigEmail>EMAIL</gitConfigEmail>
    <extensions>
      <hudson.plugins.git.extensions.impl.BuildChooserSetting>
<buildChooser class=<span class="code-quote">"com.sonyericsson.hudson.plugins.gerrit.trigger.hudsontrigger.GerritTriggerBuildChooser"</span>/>
      </hudson.plugins.git.extensions.impl.BuildChooserSetting>
      <hudson.plugins.git.extensions.impl.SubmoduleOption>
<disableSubmodules><span class="code-keyword">true</span></disableSubmodules>
<recursiveSubmodules><span class="code-keyword">false</span></recursiveSubmodules>
<trackingSubmodules><span class="code-keyword">false</span></trackingSubmodules>
<parentCredentials><span class="code-keyword">false</span></parentCredentials>
<reference/>
<timeout>10</timeout>
      </hudson.plugins.git.extensions.impl.SubmoduleOption>
    </extensions>
    <browser class=<span class="code-quote">"hudson.plugins.git.browser.GitWeb"</span>>
      <url>https:<span class="code-comment">//gerrit/gitweb?p=repo.git</url>
</span>    </browser>
  </scm>
</pre>
<p>Do you need more than the properties and scm sections? (If yes, I might have to strip some info: emails address, scripts, etc. because this example is coming from a "private" instance, I didn't have issues with OpenSource projects because they didn't do a clone recently <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> )</p>
<blockquote><p>In this case, I expect that there is a way that the job can be redefined so that it does not require the second fetch. Whether there is or not, this is a case where the second fetch is required and the plugin did not detect that the second fetch is required.</p></blockquote>
<p>In my case, the first fetch getting all branches is not needed. It doesn't really hurt but I don't need all branches.</p>

<p>But if for some reasons, all branches have to be fetched, a second fetch to the same repo is not needed if the plugin can use the first refspec in the config list instead of FETCH_HEAD. But I don't know if other users rely on FETCH_HEAD later. I guess they should use HEAD instead anyway.</p>
<blockquote><p>Sorry we missed that!</p></blockquote>
<p>All good, this happens <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Cheers,<br/>
Matt</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=398162&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398162">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matttbe" class="user-hover" rel="matttbe">matttbe</a> that should be enough.  Your freestyle job is missing the clone extension "Honor refspec on initial clone".  That option will allow the git plugin to use your defined refspec on the first fetch instead of using its default wildcard refspec.  See <a href="https://plugins.jenkins.io/git/#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/#clone-extensions</a> for the description of that option.</p>

<p>That should resolve your specific issue, but there is still a bug to fix in the git plugin because it should have detected this case and correctly realized that the second fetch is required.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matttbe" class="user-hover" rel="matttbe">matttbe</a> that should be enough.  Your freestyle job is missing the clone extension "Honor refspec on initial clone".  That option will allow the git plugin to use your defined refspec on the first fetch instead of using its default wildcard refspec.  See <a href="https://plugins.jenkins.io/git/#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/#clone-extensions</a> for the description of that option.</p>

<p>That should resolve your specific issue, but there is still a bug to fix in the git plugin because it should have detected this case and correctly realized that the second fetch is required.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matttbe.png" title="matttbe's avatar" /> matttbe:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=398164&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398164">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>Your freestyle job is missing the clone extension "Honor refspec on initial clone". That option will allow the git plugin to use your defined refspec on the first fetch instead of using its default wildcard refspec. See <a href="https://plugins.jenkins.io/git/#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/#clone-extensions</a> for the description of that option.</p></blockquote>
<p>Oh, I just found the option, I missed the Add button, more used to change settings in jenkins-job-builder <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thank you for the tip!</p>
</pre>
</details>

</ul>
</details>

<blockquote><p>Your freestyle job is missing the clone extension "Honor refspec on initial clone". That option will allow the git plugin to use your defined refspec on the first fetch instead of using its default wildcard refspec. See <a href="https://plugins.jenkins.io/git/#clone-extensions" class="external-link" target="_blank" rel="nofollow noopener">https://plugins.jenkins.io/git/#clone-extensions</a> for the description of that option.</p></blockquote>
<p>Oh, I just found the option, I missed the Add button, more used to change settings in jenkins-job-builder <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>Thank you for the tip!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=398211&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-398211">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matttbe" class="user-hover" rel="matttbe">matttbe</a> I'm not confident that I'm duplicating the exact conditions that you're seeing, but I suspect that you'll need to continue using "Preserve second fetch on checkout".</p>

<p>I've created a job that uses a parameterized refspec in the userRemoteConfig.  Unfortunately, due to a bug in the git plugin, the parameter placed in the refspec is only expanded by CLI git and not by JGit on the first fetch.  It is expanded on the second fetch.  That's a bug.  Because of that bug, you may need to continue preserving the second fetch in your Jenkins installation.</p>

<p>You don't see that bug in your current configuration because the first fetch is using the wildcard refspec rather than honoring the refspec that was set in the userRemoteConfig.  When you enable "Honor refspec in initial checkout", I expect it will fail because the parameter will not be expanded in the first call to fetch if JGit is the selected implementation.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matttbe" class="user-hover" rel="matttbe">matttbe</a> I'm not confident that I'm duplicating the exact conditions that you're seeing, but I suspect that you'll need to continue using "Preserve second fetch on checkout".</p>

<p>I've created a job that uses a parameterized refspec in the userRemoteConfig.  Unfortunately, due to a bug in the git plugin, the parameter placed in the refspec is only expanded by CLI git and not by JGit on the first fetch.  It is expanded on the second fetch.  That's a bug.  Because of that bug, you may need to continue preserving the second fetch in your Jenkins installation.</p>

<p>You don't see that bug in your current configuration because the first fetch is using the wildcard refspec rather than honoring the refspec that was set in the userRemoteConfig.  When you enable "Honor refspec in initial checkout", I expect it will fail because the parameter will not be expanded in the first call to fetch if JGit is the selected implementation.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>clabu609:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=404970&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404970">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi. We think we are encountering this problem too for our Jenkins instances and will provide details. </p>

<p>I realize that eliminating a second fetch is a nice performance optimization, but given the risks (building the wrong source code!), it would have been nice if this was a JVM system property to explicitly opt-in. Or alternatively a JVM system property to opt-out by default for all jobs. Now we have to update 1000s of Jenkins jobs (check "Preserve second fetch on checkout") to go back to the previous behavior. We don't want to risk hitting any more edge cases.</p>
</pre>
</details>

</ul>
</details>

<p>Hi. We think we are encountering this problem too for our Jenkins instances and will provide details. </p>

<p>I realize that eliminating a second fetch is a nice performance optimization, but given the risks (building the wrong source code!), it would have been nice if this was a JVM system property to explicitly opt-in. Or alternatively a JVM system property to opt-out by default for all jobs. Now we have to update 1000s of Jenkins jobs (check "Preserve second fetch on checkout") to go back to the previous behavior. We don't want to risk hitting any more edge cases.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>clabu609:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=404972&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-404972">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Oops. We discovered the old behavior can be preserved for all jobs by default via the global settings of Jenkins, so please ignore my previous comment <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Thank you for this escape hatch!</p>
</pre>
</details>

</ul>
</details>

<p>Oops. We discovered the old behavior can be preserved for all jobs by default via the global settings of Jenkins, so please ignore my previous comment <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> Thank you for this escape hatch!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-63785?focusedId=453660&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-453660">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Considering how long this has been open and how few other reports have been received, I'm closing this as "Won't fix".  In those rare cases where the second fetch is required, it can be enabled globally.  </p>
</pre>
</details>

</ul>
</details>

<p>Considering how long this has been open and how few other reports have been received, I'm closing this as "Won't fix".  In those rare cases where the second fetch is required, it can be enabled globally.  </p>
